### ISSUE TITLE
Keepalived check script timing out 

### ISSUE THREAD

**Issue by pritishmishra** [acassen/keepalived#946]
We have implemented a HA setup for our component. The track script just checks for the pid of the service and sends exit status 0 (success), if present and 1 (failure), otherwise. We poll for the track script every 6 seconds.

This is how the keepalived config looks like.
```
vrrp_script chk_sfnode {
    script "/var/vcap/jobs/keepalived/bin/get_sf_broker_status.sh"
    weight 10
    interval 6
    fall 2
    rise 2
    timeout 1
}
vrrp_instance VI_1 {
    debug 2
    interface eth0 # interface to monitor
    virtual_router_id 51 # Assign one ID for this route

    state MASTER
    priority 101 # 101 on master, 100 on backup
    unicast_src_ip 10.11.252.11 # My IP
    unicast_peer {
        10.11.253.11 # peer IP
    }

    track_script {
        chk_sfnode
    }

notify /var/vcap/jobs/keepalived/bin/notify.sh
    notify_master "/var/vcap/jobs/keepalived/bin/action.sh MASTER"
    notify_backup "/var/vcap/jobs/keepalived/bin/action.sh BACKUP"
    notify_fault  "/var/vcap/jobs/keepalived/bin/action.sh FAULT"
}

}
```

We have a test which does repeated failover of the service to check its robustness. We are observing that intermittently, at some random iteration, the check script is timing out with the following error.

`Track script chk_sfnode is being timed out, expect idle - skipping run`

When we analyse the logs for this faulty iteration, we find that the call to the track script has gone fine and the response returned by the script is also fine. However, somehow, the response didn't come back to keepalived(invoker method inside keepalived) and hence, it shows as timed out.

Once, the timeout occurs, it keeps printing the same logs and doesn't retry this operation ever, till the restart of the process.

We are using Keepalived version v1.4.2.

Has anyone faced this issue before? Are we doing something wrong? Any help will be highly appreciated. We are completely clueless about this issue.

**Comment by pqarmitage** [comment]
Could you please post the following information:

1. The keepalived logs for 1 minute before the first `Track script chk_sfnode is being timed out, expect idle - skipping run` message
2. A copy of the get_sf_broker_status.sh script
3. How long the script normally takes to run (could it be close to the 1 second timeout?)
4. How heavily loaded is the system on which keepalived is running (this will tell us if there is the likelihood of the timeout expiring).

After the first occurrence of the `Track script chk_sfnode is being timed out, expect idle - skipping run` message, have you waited for 20 minutes to see if anything changes in respect of the script (there is something in the code about a 1000 second delay if something unexpected happens)?

Could you please try the latest keepalived code from github.com/acassen/keepalived. I have just added a diagnostic message for when kill fails after a vrrp script times out, which is what I guess is happening here, and the message should tell us more about what it happening. I can see there is a potential window if keepalived times out a script but the script then terminates before keepalived attempts to kill it, but it would be helpful to have confirmation that this is the cause of your problem before attempting to fix it.

**Comment by pritishmishra** [comment]
Hi @pqarmitage, Thanks for replying to the issue. I did a bit of readup on the previous issues raised on Keepalived and found https://github.com/acassen/keepalived/issues/867 and https://github.com/acassen/keepalived/issues/864, which seemed similar to the issue, that I am facing.

I understand that this issue is fixed with the latest of keepalived version. Hence, I have now brought in the latest binary and implemented our component with it. I am re-running the tests and haven't seen any issues for the last 1 hour.

Earlier too, I had seen issues post a long time, so I am watching the issue for the day by continuing the tests. If all seems fine, then we can conclude the issue is solved with the latest version.

I will update the thread.

**Comment by pritishmishra** [comment]
We did extensive and rigorous testing, but didn't face this issue with the latest Keepalived version. So, in conclusion, this issue is fixed with the latest Keepalived release.

Thanks for your help. Closing this ticket.

**Comment by sidsagravat1406** [comment]
Which version fixed the issue?! We're using the latest stable Keepalived v2.0.7 (08/23,2018) and facing exact same issue but with additional message as follows : 

`Keepalived_vrrp[8852]: kill -15 of process /etc/keepalived/track_script_check.sh(32567) with new state 2 failed with errno 0
Keepalived_vrrp[8852]: VRRP_Script(check_service) timed_out
Keepalived_vrrp[8852]: (redis_instance) Entering FAULT STATE
`

**Comment by pqarmitage** [comment]
The checking of the return status of kill() was incorrect, and this was causing the `Keepalived_vrrp[8852]: kill -15 of process /etc/keepalived/track_script_check.sh(32567) with new state 2 failed with errno 0` message. This is fixed by commit 79f73b6.

You nevertheless appear to have a problem with /etc/keepalived/track_script_check.sh since it is timing out. You may need to increase the `timeout` parameter value in the `vrrp_script` block.
